src/stactools/hls/metadata.py

```diff
@@ -0,0 +1,148 @@
+from dataclasses import dataclass
+from datetime import datetime
+from typing import Any, Dict, List, Optional
+
+import rasterio
+from dateutil.parser import parse
+from pystac.utils import datetime_to_str
+from shapely.geometry import box, mapping
+from stactools.core.io import ReadHrefModifier
+from stactools.core.projection import reproject_geom
+
+from stactools.hls import constants, utils
+
+
+class IncorrectAssetHref(Exception):
+    """An HREF to a EO data band (not an Fmask, azimuth, or zenith band) COG is
+    required.
+    """
+
+
+@dataclass
+class Metadata:
+    """Structure to hold metadata about an HLS granule."""
+
+    id: str
+    product: str
+    version: str
+    acquisition_datetime: datetime
+    start_end_datetime: Optional[Dict[str, str]]
+    platform: str
+    instrument: List[str]
+    processing_datetime: str
+    cloud_cover: Optional[int]
+    sun_azimuth: float
+    azimuth: float
+    shape: List[int]
+    transform: List[float]
+    epsg: int
+    mgrs: Dict[str, Any]
+    geometry: Dict[str, Any]
+
+    @classmethod
+    def from_cog(
+        cls,
+        cog_href: str,
+        read_href_modifier: Optional[ReadHrefModifier] = None,
+    ) -> "Metadata":
+        """Extracts granule metadata from a COG file.
+
+        Args:
+            cog_href (str): HREF to the COG file
+            read_href_modifier (ReadHrefModifier, optional): An
+                optional function to modify the href (e.g. to add a token to a
+                url)
+
+        Returns:
+            Metadata: A Metadata dataclass
+        """
+        read_h5_href = utils.modify_href(cog_href, read_href_modifier)
+        with rasterio.open(read_h5_href) as dataset:
```

Comment:
This `read_h5_href` term is a copy-paste leftover. Edit to remove the h5 term, .e.g., make it `read_cog_href`.